Implement contrast-color( <color> max? ) CSS function

weinig opened the pull request:
#### 98862c0184b0b513c5bbd0562a5313bc6bb77f28
<pre>
Implement contrast-color( &lt;color&gt; max? ) CSS function
<a href="https://bugs.webkit.org/show_bug.cgi?id=279735">https://bugs.webkit.org/show_bug.cgi?id=279735</a>

Reviewed by Darin Adler.

Adds support for CSS Color Module 5&apos;s contrast-color() function.

Currently behind the feature flag CSSContrastColorEnabled with an
initial status of testable.

The current resolver is extremely naive, only ever choosing either
the color &apos;black&apos; or the color &apos;white&apos;, depending on which would
give the greatest contrast, which appears to be within spec, though
clearly non-ideal. Further discussion is needed to work out an
optimal algorithm for color selection.

* LayoutTests/fast/css/color-computed-contrast-color-function-expected.txt: Added.
* LayoutTests/fast/css/color-computed-contrast-color-function.html: Added.
* LayoutTests/fast/css/color-invalid-contrast-color-function-expected.txt: Added.
* LayoutTests/fast/css/color-invalid-contrast-color-function.html: Added.
* LayoutTests/fast/css/color-valid-contrast-color-function-expected.txt: Added.
* LayoutTests/fast/css/color-valid-contrast-color-function.html: Added.
    - Parsing and resolution tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new feature flag.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.
* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keyword, &apos;contrast-color&apos;.
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Add flag for contrast-color() feature flag.
* Source/WebCore/css/color/CSSContrastColorResolver.cpp: Added.
* Source/WebCore/css/color/CSSContrastColorResolver.h: Added.
    - Implements a naive resolver that just chooses white or black.

* Source/WebCore/css/color/CSSContrastColorSerialization.h: Added.
    - Generic serializer for use by either the unresolved or
      style implementation.

* Source/WebCore/css/color/CSSUnresolvedContrastColor.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedContrastColor.h: Added.
* Source/WebCore/css/color/CSSUnresolvedColor.h:
    - Adds CSSUnresolvedContrastColor variant value.

* Source/WebCore/css/color/StyleContrastColor.cpp: Added.
* Source/WebCore/css/color/StyleContrastColor.h: Added.
* Source/WebCore/css/color/StyleColor.cpp:
* Source/WebCore/css/color/StyleColor.h:
    - Add StyleContrastColor variant value. This is needed for the case
      that currentColor is used with the contrast-colors input color.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
    - Implement parsing of contrast-color().

* Source/WebCore/platform/graphics/ColorLuminance.cpp:
* Source/WebCore/platform/graphics/ColorLuminance.h:
    - Add helper to compute relative luminance from a WebCore::Color.

Canonical link: <a href="https://commits.webkit.org/283900@main">https://commits.webkit.org/283900@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0337428f16e7b17b17b0e06b547f3f8e827e7d45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20330 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71745 "Built successfully") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69809 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54872 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18638 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54181 "Passed tests") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70758 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43178 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58537 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34649 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39852 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15942 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17190 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60808 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61817 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73443 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66938 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11654 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15589 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61626 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11689 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58605 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/15034 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9508 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3126 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88706 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10300 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42876 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15647 "Found 4 new JSC stress test failures: wasm.yaml/wasm/stress/js-wasm-call-many-return-types-on-stack-no-args.js.wasm-slow-memory, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager-jettison, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43956 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43695 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/color/CSSContrastColorResolver.h:
```diff
@@ -0,0 +1,40 @@
+
+struct CSSContrastColorResolver {
+    Color color;
+    bool max;
```
Should this member be initialized? Why not have all members be initialized?

darinadler commented on Source/WebCore/css/color/CSSContrastColorResolver.h:
```diff
@@ -0,0 +1,40 @@
+
+namespace WebCore {
+
+class Color;
```
This is not needed.

darinadler commented on Source/WebCore/css/color/CSSUnresolvedContrastColor.h:
```diff
@@ -0,0 +1,54 @@
+
+struct CSSUnresolvedContrastColor {
+    UniqueRef<CSSUnresolvedColor> color;
+    bool max;
```
Same question about initialization.

darinadler commented on Source/WebCore/css/color/StyleContrastColor.h:
```diff
@@ -0,0 +1,58 @@
+
+namespace WebCore {
+
+class Color;
```
I don’t think we need this, do we?

darinadler commented on Source/WebCore/css/color/StyleContrastColor.h:
```diff
@@ -0,0 +1,58 @@
+    bool operator==(const StyleContrastColor&) const = default;
+
+    StyleColor color;
+    bool max;
```
Initialize this?

darinadler commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -111,12 +111,14 @@ CSSParserContext::CSSParserContext(const Document& document, const URL& sheetBas
 }
 
+
 void add(Hasher& hasher, const CSSParserContext& context)
```
Why add the blank line?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -661,11 +666,43 @@ static std::optional<CSSUnresolvedColor> consumeColorMixFunction(CSSParserTokenR
+    if (!color)
+        return std::nullopt;
+
+    bool max = consumeIdentRaw<CSSValueMax>(args) != std::nullopt;
```
I think `has_value` is more idiomatic?

nt1m commented on LayoutTests/fast/css/color-computed-contrast-color-function.html:
Could these be WPTs?

weinig replied:
I put a comment above in the PR about this.

darinadler replied:
https://github.com/WebKit/WebKit/pull/33841#issuecomment-2359074843

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1050,6 +1050,20 @@ CSSContentVisibilityEnabled:
+CSSContrastColorEnabled:
+  type: bool
+  category: css
+  status: testable
```
Can we enable in STP ? `preview` status

burg commented on Source/WebCore/css/color/CSSContrastColorResolver.cpp:
```diff
@@ -0,0 +1,48 @@
+
+Color resolve(const CSSContrastColorResolver& resolver)
+{
+    // FIXME: Implement support for a non-naive resolution.
```
Might be worth leaving a link to a spec issue if this is being discussed as such.

burg commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -94,6 +97,7 @@ struct ColorParserState {
 
     bool acceptQuirkyColors;
     bool colorLayersEnabled;
+    bool contrastColorEnabled;
```
Nit: member initialization

burg commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -661,11 +666,43 @@ static std::optional<CSSUnresolvedColor> consumeColorMixFunction(CSSParserTokenR
+        return std::nullopt;
+
+    auto args = consumeFunction(range);
+
```
Nit: extra newline

weinig committed:
Implement contrast-color( <color> max? ) CSS function
https://bugs.webkit.org/show_bug.cgi?id=279735

Reviewed by Darin Adler.

Adds support for CSS Color Module 5's contrast-color() function.

Currently behind the feature flag CSSContrastColorEnabled with an
initial status of testable.

The current resolver is extremely naive, only ever choosing either
the color 'black' or the color 'white', depending on which would
give the greatest contrast, which appears to be within spec, though
clearly non-ideal. Further discussion is needed to work out an
optimal algorithm for color selection.

* LayoutTests/fast/css/color-computed-contrast-color-function-expected.txt: Added.
* LayoutTests/fast/css/color-computed-contrast-color-function.html: Added.
* LayoutTests/fast/css/color-invalid-contrast-color-function-expected.txt: Added.
* LayoutTests/fast/css/color-invalid-contrast-color-function.html: Added.
* LayoutTests/fast/css/color-valid-contrast-color-function-expected.txt: Added.
* LayoutTests/fast/css/color-valid-contrast-color-function.html: Added.
    - Parsing and resolution tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new feature flag.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.
* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keyword, 'contrast-color'.
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Add flag for contrast-color() feature flag.
* Source/WebCore/css/color/CSSContrastColorResolver.cpp: Added.
* Source/WebCore/css/color/CSSContrastColorResolver.h: Added.
    - Implements a naive resolver that just chooses white or black.

* Source/WebCore/css/color/CSSContrastColorSerialization.h: Added.
    - Generic serializer for use by either the unresolved or
      style implementation.

* Source/WebCore/css/color/CSSUnresolvedContrastColor.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedContrastColor.h: Added.
* Source/WebCore/css/color/CSSUnresolvedColor.h:
    - Adds CSSUnresolvedContrastColor variant value.

* Source/WebCore/css/color/StyleContrastColor.cpp: Added.
* Source/WebCore/css/color/StyleContrastColor.h: Added.
* Source/WebCore/css/color/StyleColor.cpp:
* Source/WebCore/css/color/StyleColor.h:
    - Add StyleContrastColor variant value. This is needed for the case
      that currentColor is used with the contrast-colors input color.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
    - Implement parsing of contrast-color().

* Source/WebCore/platform/graphics/ColorLuminance.cpp:
* Source/WebCore/platform/graphics/ColorLuminance.h:
    - Add helper to compute relative luminance from a WebCore::Color.

Canonical link: https://commits.webkit.org/283900@main
